File: databasyrepo/databasyrepo/models/core/commands.py
from databasyrepo.models.core.actions import Set, Executor, Register, AppendItem, Unregister, FindAndDeleteItem, DeleteItem, InsertItem
from databasyrepo.models.core.elements import Table, Column, Index, IndexColumn
from databasyrepo.models.core.errors import IllegalCommand
from databasyrepo.models.core.reprs import Canvas, TableRepr
from databasyrepo.models.core.serializing import Serializable
from databasyrepo.models.core.validators import NodeClass, CorrectVersion, UniqueID, IfInFields
from databasyrepo.utils.validators import InvalidStateError, Always, Length, Iterable, NotEqual, Integer, Boolean, Choice

__author__ = 'Marboni'

class Command(Serializable):
    def fields(self):
        return {
            'source_version': long
        }

    def pre_validation(self, model):
        if self.__class__ not in model.commands():
            raise IllegalCommand('Command %s can not be executed on model %s.' % (self.code(), model.code()))

    def validators(self, model):
        return {
            'source_version': Always(CorrectVersion(model))
        }

    def post_validation(self, model):
        """ Custom validation logic. Starts after all validators done. Should raise IllegalCommand in case of error.
        """
        pass

    def _validate(self, model):
        for field, field_validator in self.validators(model).iteritems():
            try:
                getattr(field_validator, '__top__')
            except AttributeError:
                raise ValueError('Validator is not top-level.')
            try:
                field_validator(field, self)
            except InvalidStateError, e:
                raise IllegalCommand('Field "%s" is incorrect: %s' % (field, e.message))


    def require_checks(self):
        return True

    def _check(self, executor):
        checkers = executor.model.checkers()
        for checker in checkers:
            checker(executor).modify_errors()

    def execute(self, model):
        self.pre_validation(model)
        self._validate(model)
        self.post_validation(model)

        executor = Executor(model)
        self.do(executor)

        if self.require_checks():
            self._check(executor)

        return executor.events

    def do(self, executor):
        raise NotImplementedError


class UpdateCommand(Command):
    def __init__(self, **kwargs):
        super(UpdateCommand, self).__init__(**kwargs)

    def obj_type(self):
        """ Returns type of changeable object.
        """
        raise NotImplemented

    def obj_id_field(self):
        """ Returns name of command's field that stores object ID.
        """
        raise NotImplemented

    def obj_changeable_fields(self):
        """ Returns changeable fields list.
        """
        raise NotImplemented

    def obj_validators(self, model):
        """ Returns list of changeable fields that needs to be validated and their validators.
        """
        return {}

    def fields(self):
        fields = super(UpdateCommand, self).fields()
        fields.update({
            self.obj_id_field(): basestring,
            'fields': [basestring]
        })

        obj_proto = self.obj_type()()
        changeable_fields = self.obj_changeable_fields()
        fields_and_types = {
        field: type for field, type in obj_proto.fields().items() if field in changeable_fields
        }
        fields.update(fields_and_types)
        return fields

    def validators(self, model):
        validators = super(UpdateCommand, self).validators(model)
        obj_fields = self.obj_changeable_fields()
        obj_validators = self.obj_validators(model)

        if not set(obj_fields).issuperset(set(obj_validators.keys())):
            raise ValueError('Some of validated fields are not declared in object fields list.')

        validators.update({
            self.obj_id_field(): Always(NodeClass(model, self.obj_type())),
            'fields': Always(Iterable(min_length=1, allowed_values=obj_fields)),
        })
        for field, field_validators in obj_validators.items():
            validators[field] = IfInFields(*field_validators)
        return validators

    def do(self, executor):
        changeable_fields = self.val('fields')
        for changeable_field in changeable_fields:
            executor.execute(Set(self.val(self.obj_id_field()), changeable_field, self.val(changeable_field)))


class HistoryCommand(Command):
    def require_checks(self):
        return False

    def pre_validation(self, model):
        super(HistoryCommand, self).pre_validation(model)
        command_version = self.val('source_version')
        model_version = model.version
        if command_version != model_version:
            raise IllegalCommand(
                'Command model version %s not equals to current model version %s.' % (command_version, model_version))


class Undo(HistoryCommand):
    def pre_validation(self, model):
        super(Undo, self).pre_validation(model)
        if not model.revision_stack.can_undo():
            raise IllegalCommand('Nothing to undo.')

    def do(self, executor):
        actions = executor.model.revision_stack.undo()
        for action in actions:
            executor.execute(action)


class Redo(HistoryCommand):
    def pre_validation(self, model):
        super(Redo, self).pre_validation(model)
        if not model.revision_stack.can_redo():
            raise IllegalCommand('Nothing to redo.')

    def do(self, executor):
        actions = executor.model.revision_stack.redo()
        for action in actions:
            executor.execute(action)

# =================== TABLE ===================

class CreateTable(Command):
    def fields(self):
        fields = super(CreateTable, self).fields()
        fields.update({
            'table_id': basestring,
            'default_table_repr_id': basestring,
            'name': basestring,
            'canvas_id': basestring,
            'position': [int]
        })
        return fields

    def validators(self, model):
        validators = super(CreateTable, self).validators(model)
        validators.update({
            'table_id': Always(UniqueID(model), NotEqual('default_table_repr_id')),
            'default_table_repr_id': Always(UniqueID(model), NotEqual('table_id')),
            'name': Always(Length(1, 512)),
            'canvas_id': Always(NodeClass(model, Canvas)),
            'position': Always(Iterable(2, 2))
        })
        return validators

    def do(self, executor):
        table = Table(self.val('table_id'))
        table.set('name', self.val('name'))
        executor.execute(Register(node=table))
        executor.execute(AppendItem(field='tables', item=table))

        CreateTableRepr(
            canvas_id=self.val('canvas_id'),
            table_id=table.id,
            table_repr_id=self.val('default_table_repr_id'),
            position=self.val('position')
        ).do(executor)


class RenameTable(Command):
    def fields(self):
        fields = super(RenameTable, self).fields()
        fields.update({
            'table_id': basestring,
            'new_name': basestring
        })
        return fields

    def validators(self, model):
        validators = super(RenameTable, self).validators(model)
        validators.update({
            'table_id': Always(NodeClass(model, Table)),
            'new_name': Always(Length(1, 512))
        })
        return validators

    def do(self, executor):
        executor.execute(Set(self.val('table_id'), 'name', self.val('new_name')))


class DeleteTable(Command):
    def fields(self):
        fields = super(DeleteTable, self).fields()
        fields.update({
            'table_id': basestring,
        })
        return fields

    def validators(self, model):
        validators = super(DeleteTable, self).validators(model)
        validators.update({
            'table_id': Always(NodeClass(model, Table)),
        })
        return validators

    def do(self, executor):
        table_id = self.val('table_id')
        table = executor.model.node(table_id)

        # Delete table reprs.
        for canvas in executor.model.val_as_node('canvases', executor.model):
            for repr in canvas.val_as_node('reprs', executor.model):
                if repr.val('table').ref_id == table_id:
                    DeleteTableRepr(
                        table_repr_id=repr.id
                    ).do(executor)
                    break

        # Delete columns.
        for column_ref in table.val('columns'):
            DeleteColumn(
                column_id=column_ref.ref_id
            ).do(executor)

        # Delete indexes.
        for index_ref in table.val('indexes'):
            DeleteIndex(
                index_id=index_ref.ref_id
            ).do(executor)

        executor.execute(FindAndDeleteItem(None, 'tables', table))
        executor.execute(Unregister(table_id))


class CreateColumn(Command):
    def fields(self):
        fields = super(CreateColumn, self).fields()
        fields.update({
            'table_id': basestring,
            'column_id': basestring,
            'name': basestring,
            'type': basestring,
            'position': int
        })
        return fields

    def validators(self, model):
        validators = super(CreateColumn, self).validators(model)
        validators.update({
            'table_id': Always(NodeClass(model, Table)),
            'column_id': Always(UniqueID(model)),
            'name': Always(Length(1, 512)),
            'type': Always(Length(1, 512)),
            'position': Always(Integer(min_value=0, max_value=model.node(self.val('table_id')).items_count('columns')))
        })
        return validators

    def do(self, executor):
        table_id = self.val('table_id')
        table = executor.model.node(table_id)

        column = Column(self.val('column_id'))
        column.set('name', self.val('name'))
        column.set('type', self.val('type'))
        column.set('table', table.ref())

        executor.execute(Register(node=column))
        executor.execute(InsertItem(node_id=table_id, field='columns', index=self.val('position'), item=column))


class UpdateColumn(UpdateCommand):
    def obj_type(self):
        return Column

    def obj_id_field(self):
        return 'column_id'

    def obj_changeable_fields(self):
        return ['name', 'type', 'null', 'default']

    def obj_validators(self, model):
        return {
            'name': [Length(1, 512)],
            'type': [Length(1, 512)],
            'null': [Boolean()],
        }


class DeleteColumn(Command):
    def fields(self):
        fields = super(DeleteColumn, self).fields()
        fields.update({
            'column_id': basestring,
        })
        return fields

    def validators(self, model):
        validators = super(DeleteColumn, self).validators(model)
        validators.update({
            'column_id': Always(NodeClass(model, Column)),
        })
        return validators

    def do(self, executor):
        column_id = self.val('column_id')
        column = executor.model.node(column_id)

        table = column.val_as_node('table', executor.model)
        for index in table.val_as_node('indexes', executor.model):
            for column_index_ref in index.val('index_columns'):
                DeleteIndexColumn(index_column_id=column_index_ref.ref_id).do(executor)

        executor.execute(FindAndDeleteItem(node_id=column.val('table').ref_id, field='columns', item=column))
        executor.execute(Unregister(node_id=column_id))


class CreateIndex(Command):
    def fields(self):
        fields = super(CreateIndex, self).fields()
        fields.update({
            'table_id': basestring,
            'index_id': basestring,
            'name': basestring,
            'type': basestring,
        })
        return fields

    def validators(self, model):
        validators = super(CreateIndex, self).validators(model)
        validators.update({
            'table_id': Always(NodeClass(model, Table)),
            'index_id': Always(UniqueID(model)),
            'name': Always(Length(1, 512)),
            'type': Always(Choice(Index.TYPES)),
        })
        return validators

    def post_validation(self, model):
        super(CreateIndex, self).post_validation(model)
        if self.val('type') == Index.PRIMARY_TYPE:
            table = model.node(self.val('table_id'), Table)
            if any(index.type == Index.PRIMARY_TYPE for index in table.val_as_node('indexes', model)):
                raise IllegalCommand('Table "%s" already contains PRIMARY KEY.' % table.id)

    def do(self, executor):
        table_id = self.val('table_id')
        table = executor.model.node(table_id)

        index = Index(self.val('index_id'))
        index.set('name', self.val('name'))
        index.set('type', self.val('type'))
        index.set('table', table.ref())

        executor.execute(Register(node=index))
        executor.execute(AppendItem(node_id=table_id, field='indexes', item=index))


class UpdateIndex(UpdateCommand):
    def obj_type(self):
        return Index

    def obj_id_field(self):
        return 'index_id'

    def obj_changeable_fields(self):
        return ['name', 'type']

    def obj_validators(self, model):
        return {
            'name': [Length(1, 512)],
            'type': [Choice(Index.TYPES)],
        }

    def post_validation(self, model):
        super(UpdateIndex, self).post_validation(model)
        if self.val('type') == Index.PRIMARY_TYPE:
            index = model.node(self.val('index_id'), Index)
            table = index.val_as_node('table', model)
            for index in table.val_as_node('indexes', model):
                if index.val('type') == Index.PRIMARY_TYPE and self.val('index_id') != index.id:
                    raise IllegalCommand('Table "%s" already contains PRIMARY KEY.' % table.id)
                else:
                    break


class DeleteIndex(Command):
    def fields(self):
        fields = super(DeleteIndex, self).fields()
        fields.update({
            'index_id': basestring,
        })
        return fields

    def validators(self, model):
        validators = super(DeleteIndex, self).validators(model)
        validators.update({
            'index_id': Always(NodeClass(model, Index)),
        })
        return validators

    def do(self, executor):
        index_id = self.val('index_id')
        index = executor.model.node(index_id)

        for index_column_ref in index.val('index_columns'):
            DeleteIndexColumn(index_column_id=index_column_ref.ref_id).do(executor)

        executor.execute(FindAndDeleteItem(node_id=index.val('table').ref_id, field='indexes', item=index))
        executor.execute(Unregister(node_id=index_id))


class CreateIndexColumn(Command):
    def fields(self):
        fields = super(CreateIndexColumn, self).fields()
        fields.update({
            'index_column_id': basestring,
            'index_id': basestring,
            'column_id': basestring,
            'order': basestring,
        })
        return fields

    def validators(self, model):
        validators = super(CreateIndexColumn, self).validators(model)
        validators.update({
            'index_column_id': Always(UniqueID(model)),
            'index_id': Always(NodeClass(model, Index)),
            'column_id': Always(NodeClass(model, Column)),
            'order': Always(Choice(IndexColumn.ORDERS)),
        })
        return validators

    def do(self, executor):
        index_id = self.val('index_id')
        index = executor.model.node(index_id)

        column_id = self.val('column_id')
        column = executor.model.node(column_id)

        index_column = IndexColumn(self.val('index_column_id'))
        index_column.set('index', index.ref())
        index_column.set('column', column.ref())
        index_column.set('order', self.val('order'))

        executor.execute(Register(node=index_column))
        executor.execute(AppendItem(node_id=index_id, field='index_columns', item=index_column))


class UpdateIndexColumn(UpdateCommand):
    def obj_type(self):
        return IndexColumn

    def obj_id_field(self):
        return 'index_column_id'

    def obj_changeable_fields(self):
        return ['column', 'order']

    def obj_validators(self, model):
        return {
            'column': [NodeClass(model, Column)],
            'order': [Choice(IndexColumn.ORDERS)],
        }


class DeleteIndexColumn(Command):
    def fields(self):
        fields = super(DeleteIndexColumn, self).fields()
        fields.update({
            'index_column_id': basestring,
        })
        return fields

    def validators(self, model):
        validators = super(DeleteIndexColumn, self).validators(model)
        validators.update({
            'index_column_id': Always(NodeClass(model, IndexColumn)),
        })
        return validators

    def do(self, executor):
        index_column_id = self.val('index_column_id')
        index_column = executor.model.node(index_column_id)

        executor.execute(
            FindAndDeleteItem(node_id=index_column.val('index').ref_id, field='index_columns', item=index_column))
        executor.execute(Unregister(node_id=index_column_id))


# =================== TABLE REPR ===================

class CreateTableRepr(Command):
    def fields(self):
        fields = super(CreateTableRepr, self).fields()
        fields.update({
            'table_repr_id': basestring,
            'canvas_id': basestring,
            'table_id': basestring,
            'position': [int]
        })
        return fields

    def validators(self, model):
        validators = super(CreateTableRepr, self).validators(model)
        validators.update({
            'table_repr_id': Always(UniqueID(model)),
            'canvas_id': Always(NodeClass(model, Canvas)),
            'table_id': Always(NodeClass(model, Table)),
            'position': Always(Iterable(2, 2))
        })
        return validators

    def post_validation(self, model):
        super(CreateTableRepr, self).post_validation(model)
        canvas = model.node(self.val('canvas_id'))
        for repr in canvas.val_as_node('reprs', model):
            if repr.val('table').ref_id == self.val('table_id'):
                raise IllegalCommand('Canvas %s already has representation of table %s.' %
                                     (self.val('canvas_id'), self.val('table_id')))

    def do(self, executor):
        table = executor.model.node(self.val('table_id'), Table)
        canvas = executor.model.node(self.val('canvas_id'), Canvas)

        table_repr = TableRepr(self.val('table_repr_id'))
        table_repr.set('table', table.ref())
        table_repr.set('position', self.val('position'))
        table_repr.set('width', TableRepr.DEFAULT_REPR_WIDTH)
        executor.execute(Register(table_repr))
        executor.execute(AppendItem(canvas.id, 'reprs', table_repr))


class UpdateTableRepr(UpdateCommand):
    def obj_type(self):
        return TableRepr

    def obj_id_field(self):
        return 'table_repr_id'

    def obj_changeable_fields(self):
        return ['position', 'width']

    def obj_validators(self, model):
        return {
            'position': [Iterable(2, 2)],
            'width': [Integer(TableRepr.MIN_REPR_WIDTH)],
        }


class DeleteTableRepr(Command):
    def fields(self):
        fields = super(DeleteTableRepr, self).fields()
        fields.update({
            'table_repr_id': basestring
        })
        return fields

    def validators(self, model):
        validators = super(DeleteTableRepr, self).validators(model)
        validators.update({
            'table_repr_id': Always(NodeClass(model, TableRepr)),
        })
        return validators

    def do(self, executor):
        table_repr_id = self.val('table_repr_id')

        for canvas in executor.model.val_as_node('canvases', executor.model):
            for repr_index, repr_ref in enumerate(canvas.val('reprs')):
                if repr_ref.ref_id == table_repr_id:
                    executor.execute(DeleteItem(canvas.id, 'reprs', repr_index))
                    executor.execute(Unregister(table_repr_id))
                    return

        raise ValueError('Table repr with ID %s was not removed.' % table_repr_id)

File: databasyrepo/databasyrepo/static/repo/src/databasy/model/core/commands.js
databasy.model.core.commands.Command = databasy.model.core.serializing.Serializable.extend({
    fields: function () {
        return this._super().concat(
            'source_version'
        )
    },
    do: function (executor) {
        throw new Error('Not implemented');
    },
    execute: function (model) {
        var executor = new databasy.model.core.actions.Executor(model);
        this.do(executor);
        if (this.require_checks()) {
            this._check(executor);
        }
        return executor.events;
    },
    require_checks: function () {
        return true;
    },
    _check: function (executor) {
        var checkers = executor.model.checkers();
        for (var i = 0; i < checkers.length; i++) {
            checkers[i](executor).modify_errors();
        }
    }
});

databasy.model.core.commands.UpdateCommand = databasy.model.core.commands.Command.extend({
    obj_id_field: function () {
        throw new Error('Not implemented');
    },

    obj_changeable_fields: function () {
        throw new Error('Not implemented');
    },

    fields: function () {
        return this._super().
            concat('fields').concat(this.obj_id_field()).concat(this.obj_changeable_fields())
    },

    do: function (executor) {
        var fields = this.val('fields');
        for (var i = 0; i < fields.length; i++) {
            var field = fields[i];
            var obj_id = this.val(this.obj_id_field());
            var value = this.val(field);
            executor.execute(new databasy.model.core.actions.Set({node_id: obj_id, field: field, value: value}));
        }
    }
});


databasy.model.core.commands.HistoryCommand = databasy.model.core.commands.Command.extend({
    require_checks: function () {
        return false;
    }
});

databasy.model.core.commands.Undo = databasy.model.core.commands.HistoryCommand.extend({
    do: function (executor) {
        var actions = executor.model.revision_stack().undo();
        $.each(actions, function (i, action) {
            executor.execute(action);
        })
    }
}, {
    CODE: 'core.commands.Undo'
});

databasy.model.core.commands.Redo = databasy.model.core.commands.HistoryCommand.extend({
    do: function (executor) {
        var actions = executor.model.revision_stack().redo();
        $.each(actions, function (i, action) {
            executor.execute(action);
        })
    }
}, {
    CODE: 'core.commands.Redo'
});

databasy.model.core.commands.CreateTable = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_id',
            'default_table_repr_id',
            'name',
            'canvas_id',
            'position'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;

        var table = new core.elements.Table({
            _id: this.val('table_id'),
            name: this.val('name')
        });
        executor.execute(new core.actions.Register({node: table}));
        executor.execute(new core.actions.AppendItem({field: 'tables', item: table}));

        new core.commands.CreateTableRepr({
            table_repr_id: this.val('default_table_repr_id'),
            canvas_id: this.val('canvas_id'),
            table_id: table.id(),
            position: this.val('position')
        }).do(executor);
    }
}, {
    CODE: 'core.commands.CreateTable'
});

databasy.model.core.commands.RenameTable = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_id',
            'new_name'
        )
    },
    do: function (executor) {
        executor.execute(new databasy.model.core.actions.Set({
            node_id: this.val('table_id'),
            field: 'name',
            value: this.val('new_name')
        }));
    }
}, {
    CODE: 'core.commands.RenameTable'
});

databasy.model.core.commands.DeleteTable = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_id'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;
        var model = executor.model;

        var table_id = this.val('table_id');
        var table = model.node(table_id);

        // Delete table reprs.
        var canvases = model.val_as_node('canvases', model);
        $.each(canvases, function (index, canvas) {
            var reprs = canvas.val_as_node('reprs', model);
            //noinspection FunctionWithInconsistentReturnsJS
            $.each(reprs, function (index, repr) {
                if (table_id == repr.val('table').val('ref_id')) {
                    new core.commands.DeleteTableRepr({
                        table_repr_id: repr.id()
                    }).do(executor);
                    return false;
                }
            });
        });

        // Delete columns.
        $.each(table.val('columns'), function (i, column_ref) {
            new databasy.model.core.commands.DeleteColumn({
                column_id: column_ref.ref_id()
            }).do(executor);
        });

        // Delete indexes.
        $.each(table.val('indexes'), function (i, index_ref) {
            new databasy.model.core.commands.DeleteIndex({
                index_id: index_ref.ref_id()
            }).do(executor);
        });

        executor.execute(new core.actions.FindAndDeleteItem({field: 'tables', item: table}));
        executor.execute(new core.actions.Unregister({node_id: table_id}));
    }
}, {
    CODE: 'core.commands.DeleteTable'
});


databasy.model.core.commands.CreateColumn = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_id',
            'column_id',
            'name',
            'type',
            'position'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;

        var table = executor.model.node(this.val('table_id'), core.elements.Table);

        var column = new core.elements.Column({
            _id: this.val('column_id'),
            name: this.val('name'),
            type: this.val('type'),
            table: table.ref()
        });

        executor.execute(new core.actions.Register({node: column}));
        executor.execute(new core.actions.InsertItem({node_id: table.id(), field: 'columns', index: this.val('position'), item: column}));
    }
}, {
    CODE: 'core.commands.CreateColumn'
});

databasy.model.core.commands.UpdateColumn = databasy.model.core.commands.UpdateCommand.extend({
    obj_id_field: function () {
        return 'column_id';
    },

    obj_changeable_fields: function () {
        return [
            'name',
            'type',
            'null',
            'default'
        ]
    }
}, {
    CODE: 'core.commands.UpdateColumn'
});

databasy.model.core.commands.DeleteColumn = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'column_id'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;
        var model = executor.model;

        var column_id = this.val('column_id');
        var column = model.node(column_id);

        var table = column.val_as_node('table', executor.model);
        $.each(table.val_as_node('indexes', executor.model), function(i, index) {
            $.each(index.val('index_columns'), function(i, index_column_ref) {
                new databasy.model.core.commands.DeleteIndexColumn({
                    index_column_id:index_column_ref.ref_id()
                }).do(executor);
            });
        });

        executor.execute(new core.actions.FindAndDeleteItem({node_id: column.val('table').ref_id(), field: 'columns', item: column}));
        executor.execute(new core.actions.Unregister({node_id: column_id}));
    }
}, {
    CODE: 'core.commands.DeleteColumn'
});

databasy.model.core.commands.CreateIndex = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_id',
            'index_id',
            'name',
            'type'
        )
    },
    do: function (executor) {
        var table_id = this.val('table_id');
        var table = executor.model.node(table_id);

        var index = new databasy.model.core.elements.Index({
            _id:this.val('index_id'),
            name: this.val('name'),
            type: this.val('type'),
            table: table.ref()
        });

        executor.execute(new databasy.model.core.actions.Register({node: index}));
        executor.execute(new databasy.model.core.actions.AppendItem({node_id: table_id, field: 'indexes', item: index}));
    }
}, {
    CODE: 'core.commands.CreateIndex'
});

databasy.model.core.commands.UpdateIndex = databasy.model.core.commands.UpdateCommand.extend({
    obj_id_field: function () {
        return 'index_id';
    },

    obj_changeable_fields: function () {
        return [
            'name',
            'type'
        ]
    }
}, {
    CODE: 'core.commands.UpdateIndex'
});

databasy.model.core.commands.DeleteIndex = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'index_id'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;
        var model = executor.model;

        var index_id = this.val('index_id');
        var index = model.node(index_id);

        $.each(index.val('index_columns'), function(i, index_column_ref) {
            new databasy.model.core.commands.DeleteIndexColumn({
                index_column_id: index_column_ref.ref_id()
            }).do(executor);
        });

        executor.execute(new core.actions.FindAndDeleteItem({node_id: index.val('table').ref_id(), field: 'indexes', item: index}));
        executor.execute(new core.actions.Unregister({node_id: index_id}));
    }
}, {
    CODE: 'core.commands.DeleteIndex'
});


databasy.model.core.commands.CreateIndexColumn = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'index_column_id',
            'index_id',
            'column_id',
            'order'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;

        var index_id = this.val('index_id');
        var index = executor.model.node(index_id);

        var column_id = this.val('column_id');
        var column = executor.model.node(column_id);

        var index_column = new core.elements.IndexColumn({
            _id: this.val('index_column_id'),
            index: index.ref(),
            column: column.ref(),
            order: this.val('order')
        });

        executor.execute(new core.actions.Register({node: index_column}));
        executor.execute(new core.actions.AppendItem({node_id: index_id, field: 'index_columns', item: index_column}));
    }
}, {
    CODE: 'core.commands.CreateIndexColumn'
});


databasy.model.core.commands.UpdateIndexColumn = databasy.model.core.commands.UpdateCommand.extend({
    obj_id_field: function () {
        return 'index_column_id';
    },

    obj_changeable_fields: function () {
        return [
            'column',
            'order'
        ]
    }
}, {
    CODE: 'core.commands.UpdateIndexColumn'
});


databasy.model.core.commands.DeleteIndexColumn = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'index_column_id'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;
        var model = executor.model;

        var index_column_id = this.val('index_column_id');
        var index_column = model.node(index_column_id);

        executor.execute(new core.actions.FindAndDeleteItem({node_id: index_column.val('index').ref_id(), field: 'index_columns', item: index_column}));
        executor.execute(new core.actions.Unregister({node_id: index_column_id}));
    }
}, {
    CODE: 'core.commands.DeleteIndex'
});


databasy.model.core.commands.CreateTableRepr = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_repr_id',
            'canvas_id',
            'table_id',
            'position'
        )
    },
    do: function (executor) {
        var core = databasy.model.core;

        var table = executor.model.node(this.val('table_id'), core.elements.Table);
        var canvas = executor.model.node(this.val('canvas_id'), core.reprs.Canvas);

        var table_repr = new core.reprs.TableRepr({
            _id: this.val('table_repr_id'),
            table: table.ref(),
            position: this.val('position'),
            width: core.reprs.TableRepr.DEFAULT_REPR_WIDTH
        });
        executor.execute(new core.actions.Register({node: table_repr}));
        executor.execute(new core.actions.AppendItem({node_id: canvas.id(), field: 'reprs', item: table_repr}));
    }
}, {
    CODE: 'core.commands.CreateTableRepr'
});

databasy.model.core.commands.UpdateTableRepr = databasy.model.core.commands.UpdateCommand.extend({
    obj_id_field: function () {
        return 'table_repr_id';
    },

    obj_changeable_fields: function () {
        return ['position', 'width']
    }
}, {
    CODE: 'core.commands.UpdateTableRepr'
});

databasy.model.core.commands.DeleteTableRepr = databasy.model.core.commands.Command.extend({
    fields: function () {
        return this._super().concat(
            'table_repr_id'
        )
    },
    do: function (executor) {
        var model = executor.model;
        var table_repr_id = this.val('table_repr_id');

        var deleted = false;

        var canvases = model.val_as_node('canvases', model);
        //noinspection FunctionWithInconsistentReturnsJS
        $.each(canvases, function (index, canvas) {
            var repr_refs = canvas.val('reprs');
            //noinspection FunctionWithInconsistentReturnsJS
            $.each(repr_refs, function (index, repr_ref) {
                if (table_repr_id == repr_ref.ref_id()) {
                    executor.execute(new databasy.model.core.actions.DeleteItem({node_id: canvas.id(), field: 'reprs', index: index}));
                    executor.execute(new databasy.model.core.actions.Unregister({node_id: table_repr_id}));
                    deleted = true;
                    return false;
                }
            });
            if (deleted) {
                return false;
            }
        });

        if (!deleted) {
            throw new Error('Table repr with ID ' + table_repr_id + ' was not removed.')
        }
    }
}, {
    CODE: 'core.commands.DeleteTableRepr'
});